File: Excel/687 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/687 Fill Grid with Max Right and Down.xlsx"
input = read_excel(path, range = "A2:J11", col_names = FALSE) %>% as.matrix()
test  = read_excel(path, range = "A14:J23", col_names = FALSE) %>% as.matrix()

fill_empty = function(matrix, row, col){
  max = max(c(matrix[row, col:ncol(matrix) ], matrix[row:nrow(matrix), col]), na.rm = TRUE)
  return(max)
}
empty_cells = which(is.na(input), arr.ind = TRUE)
test_values = test[empty_cells]

filled_values = as.data.frame(empty_cells) %>%
  mutate(value = map2_dbl(row, col, ~ fill_empty(input, .x, .y))) %>%
  select(value)

all.equal(filled_values$value, test_values)
# [1] TRUE

File: Excel/600-699/687/687 Challenge.py
import pandas as pd
import numpy as np

path = "687 Fill Grid with Max Right and Down.xlsx"

input_matrix = pd.read_excel(path, usecols="A:J", skiprows=1, nrows=10, header=None).to_numpy()
test_matrix = pd.read_excel(path, usecols="A:J", skiprows=13, nrows=10, header=None).to_numpy()

def fill_empty(matrix, row, col):
    return np.nanmax(np.r_[matrix[row, col:], matrix[row:, col]])

empty_cells = np.argwhere(np.isnan(input_matrix))
filled_values = [fill_empty(input_matrix, row, col) for row, col in empty_cells]
result = np.allclose(filled_values, test_matrix[tuple(empty_cells.T)], equal_nan=True)

print(result)
